File: BikesInACampus.java
//TC: O(m*n) where m is length of workers array , n is length of bikes array
//SC: O(m*n)

//CODE:

class solution {
    public int[] assignBikes(int[][]workers , int[][]bikes){
        //null case
        if(workers == null || workers.length==0 || bikes == null || bikes.length == 0) return new int[]{};

        int m = workers.length;
        int n = bikes.length;
        HashMap<Integer, List<int []>> map = new HashMap<>();
        int min = 1000 ;   //taking min and max 
        int max = -1000;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                int[] worker = workers[i];
                int[] bike = bikes[j];
                int dist = calculateDist(worker,bike);
                min = Math.min(min,dist);
                max = Math.max(max,dist);
                if(!map.containsKey(dist)){   //if map doest not contains that distance , then add it to the map 
                    map.put(dist , new ArrayList<>());
                }
                List<int[]> li = map.get(dist); // getting the list of that distance and then adding new pair of worker and bike to list
                li.add(new int[] {i,j});
            }
        }
        int count =0;
        //now checking if worker is available to be assigned and bike is available
        boolean[] assignedWorker = new boolean[m];
        boolean[] occupiedBike = new boolean[n];
        int[] result = new int[m];
        for(int k=min;k<=max;k++){   //traversing from min distnace to max distnace
            List<int[]> li = map.get(k);
            if(li!= null){
                for(int[] wb : li){  //checking for each worker-bike pair in list
                    int w = wb[0];
                    int bi = wb[1];
                    if(!assignedWorker[w] && !occupiedBike[bi]){  //if worker and bike are availble then mark them as true and add it to the result
                        assignedWorker[w] = true;
                        occupiedBike[bi]= true;
                        result[w] =bi;
                        count++;
                    }
                    if(count == m) return result;
                }
            }
        }
        return result;
    }

    public static int calculateDist(int[] worker ; int[] bike){
        return Math.abs(worker[0]-bike[0]) + Math.abs(worker[1]- bike[1]);
    }
}

File: Bikes in a Campus.py
""""// Time Complexity : O(mn)
// Space Complexity :O(mn)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
"""
class Solution:
    def assignBikes(self, workers, bikes):
        map = {}
        m = len(workers)
        n = len(bikes)
        mini = float('inf')
        maxi = float('-inf')
        for i in range(m):
            for j in range(n):
                dist = self.calculateDistance(workers[i],bikes[j])
                mini=min(mini, dist)
                maxi=max(maxi, dist)
                if dist not in map:
                    map[dist] = []
                map[dist].append([i,j])

        assignedWorkers = [False]*m
        assignedBikes = [False]*n
        result = [0]*m
        count=0

        for i in range(mini, maxi+1, 1):
            if not map[i]:
                continue
            for k in map[i]:
                if assignedWorkers[k[0]]==False and assignedBikes[k[1]]==False:
                    assignedWorkers[k[0]] = True
                    assignedBikes[k[1]] = True
                    result[k[0]]=k[1]
                    count+=1
                    if count==len(workers):
                        return result
        return result

    def calculateDistance(self,worker,bike):

        dist = abs(worker[0]-bike[0]) + abs(worker[1]-bike[1])
        return dist

Obj=Solution()
print(Obj.assignBikes([[0,0], [1,1], [2,0]], [[1,0], [2,2], [2,1]]))